File: src/scripts/prebuild/provider.py
import argparse, os, pandas as pd, shutil, re
from pandas.core.frame import DataFrame
import unicodedata
BUILD = os.path.join('build/us')
def remove_diacritics(text):
    text = ''.join(
        c for c in unicodedata.normalize('NFD', text)
        if unicodedata.category(c) != 'Mn'
    )
    replacements = {'Å': 'A', 'Ø': 'O', 'å': 'a', 'ø': 'o', 'Æ': 'AE', 'æ': 'ae'}
    return ''.join(replacements.get(c, c) for c in text)
def sanitize(locality:str) -> str:
    #makes compatible with both filesystems and urls
    retval = locality.lower().replace(" ", "-")
    retval = re.sub(r'[^\w\-]', '', retval)
    retval = remove_diacritics(retval)
    return retval
def group_and_save_csv(df:DataFrame, output_dir, niche_):
    base_dir = output_dir
    result = {}
    group_columns = ["Region", "Locality"]
    if niche_ is None:
        group_columns.append('Niche')
    for group_keys, group in df.groupby(group_columns):
        if niche_ is None:
            region, locality, niche = group_keys
        else:
            region, locality = group_keys
            niche = niche_
        locality_clean = sanitize(locality)
        dir_path = os.path.join(base_dir, region.replace(' ', '-'), locality_clean, niche)
        create_directory = not os.path.exists(dir_path)
        if region not in result:
            result[region] = {}
        result[region][locality] = {
            "create_directory": create_directory,
            "file": os.path.join(dir_path.lower(), "content.csv"),
            "data": group.to_dict(orient="records")
        }
    return result
def execute_yaml_structure(data):
    shutil.rmtree('../build', ignore_errors=True)
    os.makedirs('../build', exist_ok=True)
    for region, localities in data.items():
        for locality, details in localities.items():
            if details["create_directory"]:
                os.makedirs(os.path.dirname(details["file"]), exist_ok=True)
            pd.DataFrame(details["data"]).to_csv(details["file"], index=False)
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('Locality')
    args = parser.parse_args()
    Locality = args.Locality
    Locality_df = pd.read_csv(Locality, dtype=str)
    yaml_data = group_and_save_csv(Locality_df, BUILD, None)
    execute_yaml_structure(yaml_data)
    

File: src/scripts/prebuild/provider.ts
import * as fs from 'fs'
import * as path from 'path'
import Papa from 'papaparse'
const BUILD = path.join('build', 'us')
export function removeDiacritics(text: string): string {
    const normalized = text.normalize('NFD').replace(/[\u0300-\u036f]/g, '')
    const replacements: Record<string, string> = {
        'Å': 'A', 'Ø': 'O', 'å': 'a', 'ø': 'o', 'Æ': 'AE', 'æ': 'ae'
    }
    return normalized.split('').map(c => replacements[c] || c).join('')
}
export function sanitize(locality: string): string {
    const cleaned = locality
        .trim()
        .toLowerCase()
        .replace(/[^a-z0-9\- ]/gi, match => removeDiacritics(match))
        .replace(/ /g, '-')
        .replace(/[^a-z0-9\-]/g, '')
    return cleaned
}
export function groupAndSaveCSV(
    data: any[],
    outputDir: string,
    niche_: string | null
): Record<string, any> {
    const result: Record<string, any> = {}
    const groupColumns = ["Region", "Locality"]
    if (niche_ === null) groupColumns.push("Niche")
    const groupedData = data.reduce((acc, row) => {
        const key = groupColumns.map(col => row[col]).join('|')
        if (!acc[key]) acc[key] = []
        acc[key].push(row)
        return acc
    }, {} as Record<string, any[]>)
    for (const groupKeys in groupedData) {
        const group = groupedData[groupKeys]
        const [region, locality, niche] = groupKeys.split('|')
        const localityClean = sanitize(locality)
        const dirPath = path.join(outputDir, region.replace(/ /g, '-'), localityClean, niche_ || niche)
        const createDirectory = !fs.existsSync(dirPath)
        if (!result[region]) result[region] = {}
        result[region][locality] = {
            createDirectory,
            file: path.join(dirPath.toLowerCase(), "content.csv"),
            data: group
        }
    }
    return result
}
export function executeYamlStructure(data: Record<string, any>): void {
    if (fs.existsSync('../build')) fs.rmSync('../build', { recursive: true })
    fs.mkdirSync('../build', { recursive: true })
    for (const region in data) {
        for (const locality in data[region]) {
            const details = data[region][locality]
            if (details.createDirectory) fs.mkdirSync(path.dirname(details.file), { recursive: true })
            const csvContent = Papa.unparse(details.data)
            fs.writeFileSync(details.file, csvContent)
        }
    }
}
if (import.meta.url === `file://${process.argv[1]}`) {
    const args = process.argv.slice(2)
    if (args.length === 0) {
        console.error('Please provide the path to the CSV file as an argument.')
        process.exit(1)
    }
    const localityPath = args[0]
    const csvContent = fs.readFileSync(localityPath, 'utf-8')
    const localityData = Papa.parse(csvContent, { header: true }).data as any[]
    const yamlData = groupAndSaveCSV(localityData, BUILD, null)
    executeYamlStructure(yamlData)
}
